pkg/rpc/context_test.go

```diff
@@ -110,6 +110,38 @@ func (*internalServer) Batch(
 	return nil, nil
 }
 
+// TestInternalServerAddress verifies that RPCContext uses AdvertiseAddr, not Addr, to
+// determine whether the apply the local server optimization.
+//
+// Prevents regression of https://github.com/cockroachdb/cockroach/issues/19991.
+func TestInternalServerAddress(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	stopper := stop.NewStopper()
+	defer stopper.Stop(context.TODO())
+
+	// Can't be zero because that'd be an empty offset.
+	clock := hlc.NewClock(timeutil.Unix(0, 1).UnixNano, time.Nanosecond)
+
+	baseCtx := testutils.NewNodeTestBaseContext()
+
+	baseCtx.Addr = "127.0.0.1:9999"
+	baseCtx.AdvertiseAddr = "127.0.0.1:8888"
+
+	serverCtx := NewContext(log.AmbientContext{Tracer: tracing.NewTracer()}, baseCtx, clock, stopper)
+
+	internal := &internalServer{}
+	serverCtx.SetLocalInternalServer(internal)
+
+	if is := serverCtx.GetLocalInternalServerForAddr(baseCtx.Addr); is != nil {
```

Comment:
It strikes me that this test case is kind of the opposite of what we want. We don't actually expect `baseCtx.Addr` to be the target of any traffic, but if it was I don't know why we wouldn't want to use the local server optimization for it.

More of an observation than anything, though, given that we don't expect it to be the target of real traffic.